cloud/amazon/ec2_vpc_dhcp_options.py

```diff
@@ -128,105 +157,204 @@
+
+    # First check if we were given a dhcp_options_id
+    if not params['dhcp_options_id']:
+    # No, so create new_options from the parameters
+        if params['dns_servers'] != None:
+            new_options['domain-name-servers'] = params['dns_servers']
+        if params['netbios_name_servers'] != None:
+            new_options['netbios-name-servers'] = params['netbios_name_servers']
+        if params['ntp_servers'] != None:
+            new_options['ntp-servers'] = params['ntp_servers']
+        if params['domain_name'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['domain-name'] = [ params['domain_name'] ]
+        if params['netbios_node_type'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['netbios-node-type'] = [ str(params['netbios_node_type']) ]
+    # If we were given a vpc_id then we need to look at the options on that
+    if params['vpc_id']:
+        existing_options = fetch_dhcp_options_for_vpc(connection, params['vpc_id'])
+        # if we've been asked to inherit existing options, do that now
+        if params['inherit_existing']:
+            for option in [ 'domain-name-servers', 'netbios-name-servers', 'ntp-servers', 'domain-name', 'netbios-node-type']:
+                if existing_options.options.get(option) and (not new_options[option] or [''] == new_options[option]):
+                    new_options[option] = existing_options.options.get(option)
+
+        # Do the vpc's dhcp options already match what we're asked for? if so we are done
+        if existing_options and new_options == existing_options.options:
+            found = True
+            dhcp_option = existing_options
+            module.exit_json(changed=changed, new_options=new_options, dhcp_options_id=existing_options.id)
+
+        # If no vpc_id was given, or the options don't match then look for an existing set using tags
+        found, dhcp_option = match_dhcp_options(connection, params['tags'], new_options)
+
+    # Now let's cover the case where there are existing options that we were told about by id
+    # If a dhcp_options_id was supplied we don't look at options inside, just set tags (if given)
+    else:
+        supplied_options = connection.get_all_dhcp_options(dhcp_options_ids=params['dhcp_options_id'])
+        found = True
+        dhcp_option = supplied_options[0]
+        if params['state'] != 'absent' and options['tags']:
+            ensure_tags(connection, options['tags'], False, module.check_mode)
+
+    # Now we have the dhcp options set, let's do the necessary
+
+    # if we found options we were asked to remove then try to do so
+    if found and params['state'] == 'absent' and not module.check_mode:
+        changed = remove_dhcp_options_by_id(connection, dhcp_option.id)
+        module.exit_json(changed=changed, new_options={})
```

Comment:
I think your handling of module.check_mode here is off. You should do something like:

```python
if found and params['state'] == 'absent':
    if not module.check_mode:
        changed = remove_dhcp_options_by_id(connection, dhcp_option.id)
    module.exit_json(changed=changed, new_options={})
````